src/coprocessor/gadgets.rs

```diff
@@ -124,22 +134,364 @@ pub fn construct_list<F: LurkField, CS: ConstraintSystem<F>>(
+
+    let hash = hash_poseidon(
+        &mut cs.namespace(|| "hash"),
+        vec![
+            a.tag().clone(),
+            a.hash().clone(),
+            b.tag().clone(),
+            b.hash().clone(),
+        ],
+        store.poseidon_cache.constants.c4(),
+    )?;
+
+    implies_equal(
+        &mut cs.namespace(|| "hash equality"),
+        not_dummy,
+        tuple.hash(),
+        &hash,
+    );
+
+    Ok((a, b))
+}
+
+/// Deconstructs an `AllocatedPtr` that's assumed to be a composition of three others
+pub fn deconstruct_tuple3<F: LurkField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    store: &Store<F>,
+    not_dummy: &Boolean,
+    tuple: &AllocatedPtr<F>,
+) -> Result<(AllocatedPtr<F>, AllocatedPtr<F>, AllocatedPtr<F>), SynthesisError> {
+    let (a, b, c) = if not_dummy.get_value() == Some(true) {
+        let idx = get_ptr(tuple, store)?.get_index3().expect("invalid Ptr");
+        let (a, b, c) = store.expect_3_ptrs(idx);
+        (store.hash_ptr(a), store.hash_ptr(b), store.hash_ptr(c))
+    } else {
+        (ZPtr::dummy(), ZPtr::dummy(), ZPtr::dummy())
+    };
+
+    let a = AllocatedPtr::alloc(&mut cs.namespace(|| "fst"), || Ok(a))?;
+    let b = AllocatedPtr::alloc(&mut cs.namespace(|| "snd"), || Ok(b))?;
+    let c = AllocatedPtr::alloc(&mut cs.namespace(|| "trd"), || Ok(c))?;
+
+    let hash = hash_poseidon(
+        &mut cs.namespace(|| "hash"),
+        vec![
+            a.tag().clone(),
+            a.hash().clone(),
+            b.tag().clone(),
+            b.hash().clone(),
+            c.tag().clone(),
+            c.hash().clone(),
+        ],
+        store.poseidon_cache.constants.c6(),
+    )?;
+
+    implies_equal(
+        &mut cs.namespace(|| "hash equality"),
+        not_dummy,
+        tuple.hash(),
+        &hash,
+    );
+
+    Ok((a, b, c))
+}
+
+/// Deconstructs an `AllocatedPtr` that's assumed to be a composition of four others
+pub fn deconstruct_tuple4<F: LurkField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    store: &Store<F>,
+    not_dummy: &Boolean,
+    tuple: &AllocatedPtr<F>,
+) -> Result<
+    (
+        AllocatedPtr<F>,
+        AllocatedPtr<F>,
+        AllocatedPtr<F>,
+        AllocatedPtr<F>,
+    ),
+    SynthesisError,
+> {
+    let (a, b, c, d) = if not_dummy.get_value() == Some(true) {
+        let idx = get_ptr(tuple, store)?.get_index4().expect("invalid Ptr");
+        let (a, b, c, d) = store.expect_4_ptrs(idx);
+        (
+            store.hash_ptr(a),
+            store.hash_ptr(b),
+            store.hash_ptr(c),
+            store.hash_ptr(d),
+        )
+    } else {
+        (ZPtr::dummy(), ZPtr::dummy(), ZPtr::dummy(), ZPtr::dummy())
+    };
+
+    let a = AllocatedPtr::alloc(&mut cs.namespace(|| "fst"), || Ok(a))?;
+    let b = AllocatedPtr::alloc(&mut cs.namespace(|| "snd"), || Ok(b))?;
+    let c = AllocatedPtr::alloc(&mut cs.namespace(|| "trd"), || Ok(c))?;
+    let d = AllocatedPtr::alloc(&mut cs.namespace(|| "frt"), || Ok(d))?;
```

Comment:
This is the point at which I break down and note that this namespacing scheme isn't great. I would just use a, b, c, d (and same for all the above examples as well).